assistant tests

In [1]:
from openai import OpenAI
from openai import types
import os
import json
from dotenv import load_dotenv

In [2]:
load_dotenv('../.env')
token = os.getenv('TOKEN_API_GPT')
tokenP = os.getenv('TOKEN_API_PER')
clientP = OpenAI(api_key='', base_url="https://api.perplexity.ai")
client = OpenAI(api_key='')

In [11]:
assistant = client.beta.assistants.retrieve('asst_gx0OWMBDg3kA2pkHyUHIGTJs')

In [45]:
thread = client.beta.threads.create()

In [46]:
thread.id

'thread_lHNHRwPN8Txd4XPgEkhIKrh4'

In [47]:
thread1 = client.beta.threads.retrieve(thread.id)

In [48]:
thread1

Thread(id='thread_lHNHRwPN8Txd4XPgEkhIKrh4', created_at=1726321055, metadata={}, object='thread', tool_resources=ToolResources(code_interpreter=ToolResourcesCodeInterpreter(file_ids=[]), file_search=None))

In [37]:
thread = thread.to_dict()

In [38]:
thread

{'id': 'thread_l4FW4JSEtWaZGvL5NEnLgnov',
 'created_at': 1726320147,
 'metadata': {},
 'object': 'thread',
 'tool_resources': {}}

In [44]:
types.beta.thread.Thread(thread)

TypeError: 'module' object is not callable

In [53]:
thread.id, thread1.id

('thread_lHNHRwPN8Txd4XPgEkhIKrh4', 'thread_lHNHRwPN8Txd4XPgEkhIKrh4')

In [57]:
message = client.beta.threads.messages.create(
  thread_id=thread1.id,
  role="user",
  content="hehehehe"
)

In [60]:
run = client.beta.threads.runs.create_and_poll(
  thread_id=thread.id,
  assistant_id='asst_gx0OWMBDg3kA2pkHyUHIGTJs',
)

In [3]:
messages = client.beta.threads.messages.list(
    thread_id='thread_QotnvBfr7tZoAT7ko8rafPYM'
  )
messages.to_dict()

{'data': [{'id': 'msg_nY2aA2Q4YdpYG1hgWrv7kJpp',
   'assistant_id': 'asst_gx0OWMBDg3kA2pkHyUHIGTJs',
   'attachments': [],
   'content': [{'text': {'annotations': [],
      'value': 'Готово\n\nИлья\nКрамник\ninfo.einfluence@gmail.com\n89200478047\nЁ-Инфлюенс\nCEO\nЁ-Инфлюенс\nрекламное агентство по продвижению бизнесов через микроблогеров. Осуществляет полный комплекс услуг: подбор блогеров, анализ блогеров, переговоры с блогерами, создание креативов, маркировка рекламы, создание UTM меток, сопровождение публикаций. Акцент на работу с микроблогерами в Telegram и Instagram.\nБизнес хочет делегировать операционную работу, связанную с подбором и переговорами с десятками блогеров, и весь процесс продвижения продукта.\nB2C сегмент, эдтех, мода, красота, продажа на маркетплейсах товаров стоимостью до 2000 рублей\nот 7 сотрудников\nот 2 лет\nот 40 миллионов\nCEO, CMO'},
     'type': 'text'}],
   'created_at': 1729508984,
   'metadata': {},
   'object': 'thread.message',
   'role': 'assistant'

In [63]:
messages.data[0].content[0].text.value

'Понимаю, что это может занять некоторое время. Если вам удобнее, вы можете ответить голосовым сообщением. Расскажите, пожалуйста, немного больше о вашем продукте: как он называется, какие задачи решает для ваших клиентов и какие ключевые преимущества и ценность он предлагает?'

In [24]:
if run.status == 'completed': 
  messages = client.beta.threads.messages.list(
    thread_id=thread.id
  )
  print(messages.to_dict()['data'][0]['content'][0]['text']['value'])
else:
  print(run.status)

Рада познакомиться, Тимофей! Могу я уточнить, в какой компании вы работаете и какая у вас должность?


In [75]:
thread = client.beta.threads.create()

In [84]:
text = f"""
Твоя задача написать письмо для компании ДАТАКРАТ-Е, ООО НПФ, ее область деятельности 62.09 и инн - 6663041348, а сайт www.datakrat.ru
Их выручка за прошлый год - 694293000 рублей, а количество сотрудников - 66
Ты пишешь письмо от лица человека - его зовут Иван Иванов, он работает в AI Solutions и занимает должность менеджер по продажам, а его почта - ivanov.sales@company.com
Твоя задача продать продукт - AI Sales Assistant, вот его описание Мы предлагаем продукт AI Sales Assistant, который позволяет автоматизировать взаимодействие с клиентами, значительно ускорять процессы коммуникации и облегчать рутинные задачи, такие как ведение баз данных, обработка запросов и организация встреч. Продукт также помогает глубже анализировать поведение клиентов и находить оптимальные точки взаимодействия, делая процесс продаж более эффективным и персонализированным. и проблемы, которые он решает Сокращение времени на обработку запросов, Увеличение коэффициента конверсии, Автоматизация рутинных задач
"""

In [85]:
def analyze_website(url: str):
    messages = [
        {
            "role": "system",
            "content": (
                "You are an artificial intelligence assistant and you need to "
                "engage in a helpful, detailed, polite conversation with a user."
            ),
        },
        {
            "role": "user",
            "content": (
                f"Получи информацию об описание компании, описание их основного продукта/услуги, последние новости за последний месяц {url}"
            ),
        },
    ]
    
    response = clientP.chat.completions.create(
        model="llama-3.1-sonar-large-128k-online",
        messages=messages,
    )
    return response.choices[0].message.content

    # response_stream = clientP.chat.completions.create(
    #     model="llama-3.1-sonar-large-128k-online",
    #     messages=messages,
    #     stream=True,
    # )
    # for response in response_stream:
        # print(response)

In [67]:
analyze_website('http://www.datakrat.ru')

'### Описание Компании\n\nКомпания ДатаКрат является одним из ведущих разработчиков программного обеспечения и системных интеграторов в области автоматической идентификации и автоматизации торговли в России.\n\n- Компания была основана 12 марта 1992 года в Екатеринбурге на базе лаборатории факультета математики и механики Уральского государственного университета.\n- ДатаКрат имеет широкую географическую присутствие, с центральным офисом в Екатеринбурге и филиалами в многих крупных городах России, включая Челябинск, Тюмень, Уфу, Пермь, и другие.\n- Компанияemploys более 500 сотрудников в 22 офисах по всей стране.\n\n### Основные Продукты и Услуги\n\nДатаКрат предлагает широкий спектр продуктов и услуг, включая:\n\n- **Автоматическая идентификация**: Решения по штриховому кодированию, автоматической идентификации и маркировке товаров.\n- **Автоматизация торговли**: Комплексная автоматизация торговых предприятий, включая системы управления торговлей, кассы самообслуживания, системы операт

In [86]:
thread = client.beta.threads.create()

message = client.beta.threads.messages.create(
  thread_id=thread.id,
  role="user",
  content=text
)
run = client.beta.threads.runs.create_and_poll(
  thread_id=thread.id,
  assistant_id='asst_Ag8SRhkXXleq6kgdW0zWtkAP',
    
)

while run.status != "completed":
    run = client.beta.threads.runs.retrieve(thread_id=thread.id, run_id=run.id)
    if run.status == "requires_action":
        for tool_call in run.required_action.submit_tool_outputs.tool_calls:
            if tool_call.function.name == "analyze_website":
                url = json.loads(tool_call.function.arguments)['url']
                analysis_result = analyze_website(url)
                client.beta.threads.runs.submit_tool_outputs(
                    thread_id=thread.id,
                    run_id=run.id,
                    tool_outputs=[{
                        "tool_call_id": tool_call.id,
                        "output": analysis_result
                    }]
                )

messages = client.beta.threads.messages.list(
    thread_id=thread.id
  )
messages.data[0].content[0].text.value

'Превью: \nООО НПФ "ДАТАКРАТ-Е" (ИНН: 6663041348) специализируется на автоматической идентификации и системной интеграции (ОКВЭД 62.09). Компания разрабатывает и внедряет системы автоматической идентификации и автоматизации для торговых предприятий и промышленности, а также занимается установкой и обслуживанием систем безопасности. В компании работают 66 сотрудников, а годовая выручка за прошлый год составила 694 293 000 рублей. Сайт компании: www.datakrat.ru\n\nТема: Запрос на контакты технического директора ООО НПФ "ДАТАКРАТ-Е"\n\nПисьмо:\n\nДобрый день,\n\nЯ Иван Иванов из AI Solutions. Ваша работа в области автоматической идентификации и системной интеграции действительно впечатляет! Ваше стремление к инновациям и широта географического присутствия делают вас лидером в этой области.\n\nМы предлагаем AI Sales Assistant — продукт, который автоматизирует взаимодействие с клиентами и помогает ускорить процессы коммуникации. Это решение может быть полезным для оптимизации ваших внутренн

Превью: ООО НПФ "ДАТАКРАТ-Е" (ИНН: 6663041348) является ведущим разработчиком программного обеспечения и системных интегратором в области автоматической идентификации и автоматизации торговли. Компания предлагает решения для автоматизации торговых и промышленных предприятий. В компании работают 66 сотрудников, а годовая выручка за прошлый год составила 694 293 000 рублей. Сайт компании: datakrat.ru\n\nТема: Запрос на контакты технического директора «ДАТАКРАТ-Е»\n\nПисьмо:\n\nДобрый день,\n\nМеня зовут Иван Иванов, я представляю компанию AI Solutions. Ваша компания «ДАТАКРАТ-Е» впечатляет своими решениями в области автоматической идентификации и автоматизации торговли. Это действительно вдохновляющий пример успешной интеграции технологий в бизнес-процессы.\n\nМы разработали продукт AI Sales Assistant, который автоматизирует взаимодействие с клиентами, ускоряет процессы коммуникации и облегчает рутинные задачи. Это может помочь вашей команде глубже анализировать поведение клиентов и оптимизировать точки взаимодействия, делая процесс продаж более эффективным.\n\nБуду признателен, если вы подскажете, с кем из вашей команды, возможно, с техническим директором, я мог бы обсудить потенциальное сотрудничество и внедрение нашего решения?\n\nС уважением,  \nИван Иванов  \nМенеджер по продажам, AI Solutions  \nivanov.sales@company.com

In [10]:
thread.id

'thread_xIwQXuYLcpGrrHI5RnTc1gWR'

In [130]:
messages = client.beta.threads.messages.list(
    thread_id='thread_Nvjcc2GU7mXJ0lpG17UIj7GS'
  )
messages.data[0].content[0].text.value

'```json\n{\n  "name": "Иван",\n  "surname": "Иванов",\n  "email": "ivanov.sales@company.com",\n  "tel": "+7 (123) 456-78-90",\n  "company_name": "AI Solutions",\n  "jobtitle": "Менеджер по продажам",\n  "product_name": "AI Sales Assistant",\n  "product_description": "интеллектуальное решение для автоматизации процесса продаж и взаимодействия с клиентами",\n  "problem_solved": "сокращение времени на обработку заявок, повышение конверсии, автоматизация рутинных задач",\n  "target_okveds": ["62.01", "64.92", "47.91", "85.41", "69.10", "47.00"],\n  "target_number_employees": [100, 1000],\n  "target_number_years_existence": [5, 15],\n  "target_revenue_last_year": [500000000, 5000000000],\n  "target_jobtitle": ["директор по продажам", "коммерческий директор", "руководитель отдела продаж", "маркетинг-директор", "руководитель отдела маркетинга", "IT-директор", "руководитель отдела IT", "генеральный директор"]\n}\n```'

In [129]:
import json
data = messages.data[0].content[0].text.value
if data[8] == '{' and data[-5] == '}':
  json.loads(data[8:-6])
else: json.loads(data)

JSONDecodeError: Expecting ',' delimiter: line 15 column 222 (char 891)

In [18]:
text1 = """
Вы — AI-ассистент, специализирующийся на создании email outreach писем, которые направляются в компании с целью перенаправления их нужному сотруднику. Письма должны представлять ваш продукт или услугу, используя партнерский тон, который делает их предложением сотрудничества, а не рекламной кампанией.

Контекст и Информация:

Компания:

Название: {company.company_name}
Отрасль: {company.okveds}
ИНН: {company.inn}
Сайт: {company.site}, когда получаешь ссылку на сайт обязательно используй инструмент/функцию analyze_website. Чтобы получить информацию об описание компании, описание их основного продукта/услуги, последние новости за последний месяц

Последняя годовая выручка: {company.revenue_last_year}
Численность сотрудников: {company.number_employees}

Отправитель:

Имя: {user.name} {user.surname}
Компания: {user.company_name}
Должность: {user.jobtitle}
Эл. почта: {user.email}

Продукт:

Название продукта: {user.product_name}
Описание: {user.product_description}
Решаемые проблемы: {user.problem_solved}

Контактное лицо:

Позиция, которая пользователя интересует в компании: [company.job title]

Ключевые навыки:

Партнёрский тон: Используйте язык, который делает акцент на сотрудничестве и общих интересах, избегая рекламных клише.

Персонализация контекста компании: Включите уникальные данные и достижения компании, чтобы продемонстрировать осведомлённость и уважение.

Целевая структура: Обеспечьте логическую последовательность: приветствие, похвалить продукт компании (отметить, что ты знаешь его  и компанию) , рассказать о себе и своем продукте, но без особой рекламы,  и просьба о перенаправлении к подходящему лицу, например, к техническому директору. Следовать примерам писем, их структуре, которые приведены ниже.

Релевантность и ценность: Подчеркните, как ваш продукт может быть полезен и актуален для поддержания текущих инициатив компании.

Эффективные примеры и шаблоны: Применяйте лучшие практики, чтобы письмо выглядело профессионально и побуждало к положительному действию. Примеры можно увидеть ниже.

Требования к письму:

Формат: 3-5 абзацев по 2-3 предложения каждый.
Главная цель: Установить контакт, представив продукт как часть партнёрского предложения.
Персонализация: Уникализируйте письмо, избегайте шаблонности.
Источники информации: Не упоминайте, откуда получена информация о компании.
Фокус на диалог: Способствуйте созданию возможностей для продолжения общения и сотрудничества.
Тон: Пиши как человек - просто, но при этом сохраняя логику и факты. Сам ничего лишнего не придумывай.

Примеры Превью:

1. Пример первый: 
ООО “Доверенная Среда” - «Триафлай» (ИНН: 7701081120) специализируется на разработке российских BI-решений в сфере информационных технологий (ОКВЭД 62.01). Основное направление деятельности компании – создание BI-платформы для систематизации и автоматизации процессов сбора, обработки, анализа и хранения больших объемов данных. Решения “Триафлай” помогают визуализировать данные, строить отчеты и поддерживать принятие стратегических решений на основе моделирования и прогнозирования. В компании работают 56 сотрудников, а годовая выручка за прошлый год составила 254 662 000 рублей. Сайт компании: triafly.ru

2. Пример второй: 
ООО “Технологические Ресурсы” - «ТехРесурс» (ИНН: 7705123456) занимается разработкой программного обеспечения для автоматизации производства (ОКВЭД 62.09). Основное направление деятельности компании – создание комплексных решений для цифровой трансформации производственных предприятий. Платформы “ТехРесурс” помогают оптимизировать логистику, производственные процессы и управление ресурсами. В компании работают 134 сотрудника, а годовая выручка за прошлый год составила 530 970 000 рублей. Сайт компании: techresource.ru

3. Пример третий: 
ООО “Генподряд” - «СтройМастер» (ИНН: 7708123456) осуществляет общестроительные работы и реконструкцию объектов (ОКВЭД 41.10). Основное направление деятельности компании – генподряд на строительство жилых и административных зданий, а также реставрационные работы и модернизация объектов инфраструктуры. В компании работают 312 сотрудников, а годовая выручка за прошлый год составила 1 123 650 000 рублей. Сайт компании: stroymaster.ru

4. Пример четвертый: ООО “СтройИнвест” - «СтройПрофиль» (ИНН: 7724098765) занимается проектированием и строительством жилых и коммерческих объектов (ОКВЭД 41.20). Основное направление деятельности компании – выполнение полного цикла строительных работ, от разработки проектов до их реализации. Компания специализируется на строительстве жилых комплексов и коммерческих центров. В компании работают 245 сотрудников, а годовая выручка за прошлый год составила 987 340 000 рублей. Сайт компании: stroyprofil.ru

5. Пример пятый: ООО “Рекламные Технологии” - «МаркетМедиа» (ИНН: 7709112234) предоставляет услуги в области цифрового маркетинга и рекламных кампаний (ОКВЭД 73.11). Основное направление деятельности компании – разработка стратегий интернет-рекламы, управление контекстной и медийной рекламой, а также продвижение брендов через социальные сети и SEO. В компании работают 45 сотрудников, а годовая выручка за прошлый год составила 112 540 000 рублей. Сайт компании: marketmedia.ru

Примеры Тем для Писем:

1. "Запрос на контакты HR-директора ООО “ИнфоСистемы”"
2. "Запрос на контакты технического директора Триафлай"
3. "Поделитесь контактами технического директора «ТехРесурс»"
4. "Запрос на контакты директора по ИТ ООО “Рекламные Стратегии”"
5. "Ищем контакты директора по продукту «СтройСфера» для обсуждения проекта"

Примеры Писем:

1. Пример первый:
Добрый день,

Я Владислав Богдан из Flowgen. Ваши инновации в облачном майнинге с Cyberian Mine просто впечатляют! Вы действительно упрощаете доступ к майнингу и постоянно улучшаете условия для своих клиентов.

У нас есть платформа, которая объединяет ИИ инструменты и помогает повысить эффективность сотрудников на 42%. Это может стать отличным дополнением к вашим внутренним процессам.

Можете подсказать, с кем из вашей команды, возможно, с HR-директором, я мог бы обсудить возможности сотрудничества и внедрение нашего решения?

С уважением,
Владислав Богдан
CEO, Flowgen
vlbog@flowgen.ru


2. Пример второй:
Добрый день,

Я Владислав Богдан из Flowgen. Ваши замороженные полуфабрикаты от ООО "ДОКА ВКУСА" — просто находка! Вы действительно создаете продукцию, которая ценится за свое качество.

Мы разработали платформу, объединяющую ИИ инструменты, которая помогает повысить эффективность сотрудников на 42%. Это может быть именно тем решением, которое дополнит ваши процессы.

Можете подсказать, с кем из вашей команды, возможно, из отдела информационных технологий или инноваций, я мог бы обсудить потенциальное сотрудничество и внедрение нашего решения?

С уважением,
Владислав Богдан
CEO, Flowgen
vlbog@flowgen.ru

3. Пример третий: 
Добрый день,

Я Егор Летов из компании Ledokol IT. Ваши наушники Neiry Buds+ — это что-то невероятное! Такие технологии в области нейротехнологий впечатляют и действительно делают жизнь лучше. Обязательно присмотрю их для себя))

Мы занимаемся разработкой ИИ-решений, которые могут упростить работу и повысить эффективность ваших процессов и сотрудников.

С кем из вашей команды, возможно, с техническим директором, я мог бы обсудить, как мы можем вместе поработать и обсудить потенциальное сотрудничество?

С уважением,
Егор Летов
Генеральный директор, Ledokol IT
vlbog@flowgen.ru


Формат ответа:

Превью: [превью компании]
Тема: [тема письма]
Письмо: [текст письма]
"""

In [22]:
# Функция для анализа сайта
def analyze_website():
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=[
            {"role": "system", "content": text1},
            {"role": "user", "content": text}
        ],
        functions=[
            {
                "name": "analyze_website",
                "description": "Automatically analyzes a website and provides a comprehensive summary of the company description, their main product/service, and the latest news from the last month",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "url": {
                            "type": "string",
                            "description": "The URL of the website to analyze"
                        }
                    },
                    "required": ["url"]
                }
            }
        ],
        function_call={"name": "analyze_website", "arguments": f'{{"url": "www.datakrat.ru"}}'}
    )

    data = response.choices[0].message.content
    return data

result = analyze_website()
print(result)

None
